cmd/log/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+
+    let mut update_stringbuf = || -> Stringbuf {
+        let _info = core.op_start().unwrap();
+        core.read_8(ringbuf_var.addr, buf.as_mut_slice()).unwrap();
+        core.op_done().unwrap();
+
+        // Load the stringbuf struct with data from buf
+        let ringbuf_val: Value = Value::Struct(
+            reflect::load_struct(hubris, &buf, definition, 0).unwrap(),
+        );
+
+        let cell: StaticCell = StaticCell::from_value(&ringbuf_val).unwrap();
+        Stringbuf::from_value(&cell.cell.value).unwrap()
+    };
+
+    // use a BTree to keep a sorted hashmap of the recent log entries,  this prevents reprinting log
+    // lines as well as keeping them ordered.
+    let mut full_log = BTreeMap::new();
+
+    //
+    // TODO
+    // set ctrl c handler before looping
+    // this will ensure that op_done is always called even if interrupted
+    // see jira https://rivosinc.atlassian.net/browse/SW-440
+    //
+
+    loop {
+        //
+        // Read the log from the core and convert to a big string
+        //
+        let stringbuf = update_stringbuf();
+        let buffer = stringbuf.buffer.as_slice();
+        let log = str::from_utf8(buffer);
+        if let Err(_err) = log {
+            println!("Invalid Log: {:?}", stringbuf);
+        }
+
+        //
+        // split log file up into lines based on null delimeter,
+        // also count the characters
+        //
+        let log = log.unwrap();
+        let mut lines: Vec<&str> = log.split('\0').collect();
+        let char_counts: Vec<usize> =
+            lines.iter().map(|line| line.len()).collect();
+
+        //
+        // remove the line that is currently being overwritten
+        //
+        let last_idx: usize = stringbuf.last.unwrap() as usize;
+        let last_char = buffer[last_idx];
+        if last_char != 0 && last_char != b'\r' {
+            let mut total_count = 0;
+            for (char_count, (i, _line)) in
+                zip(char_counts, lines.clone().iter().enumerate())
+            {
+                if total_count <= last_idx
+                    && total_count + char_count > last_idx
+                {
+                    lines.swap_remove(i);
+                    break;
+                }
+                total_count += char_count;
+            }
+        }
+
+        for (i, l) in lines.iter().enumerate() {
+            //
+            // extract and verify timestamp is complete and does not contain letters
+            //
+            let timestamp = extract_timestamp(l);
+            if let Err(_err) = timestamp {
+                continue;
+            }
+            let timestamp: u64 = timestamp.unwrap();
+
+            //
+            // Insert line into map
+            //
+            if let std::collections::btree_map::Entry::Vacant(_e) =
+                full_log.entry(timestamp)
+            {
+                // the first line could contain a partial timestamp, so verify it starts with a
```

Comment:
Why is there a special check for `i = 0` containing a partial timestamp? Couldn't this be true for any entry?